networking_aci/plugins/ml2/drivers/mech_aci/common.py

```diff
@@ -135,11 +137,95 @@ def get_hosts_on_network(self, context, network_id, level=None, with_segment=Fal
 
+    def get_ports_on_network_by_physnet_prefix(self, context, network_id, physical_network_prefix):
+        # get all ports for a network that are on a segment with a physnet prefix
+        fields = [
+            segment_models.NetworkSegment.id, segment_models.NetworkSegment.physical_network,
+            models_v2.Port.id, models_v2.Port.project_id
+        ]
+        query = context.session.query(*fields)
+        query = query.filter(models_v2.Port.network_id == network_id)
+        query = query.join(ml2_models.PortBindingLevel, ml2_models.PortBindingLevel.port_id == models_v2.Port.id)
+        query = query.join(segment_models.NetworkSegment,
+                           ml2_models.PortBindingLevel.segment_id == segment_models.NetworkSegment.id)
+        query = query.filter(segment_models.NetworkSegment.physical_network.like('{}%'.format(physical_network_prefix)))
+
+        result = []
+        for entry in query.all():
+            result.append({
+                'segment_id': entry[0],
+                'port_id': entry[1],
+                'physical_network': entry[2],
+                'project_id': entry[3],
+            })
+
+        return result
+
+    def get_bound_projects_by_physnet_prefix(self, context, physical_network_prefix):
+        # get all projects that have a port bound to a segment with this prefix
+        query = context.session.query(models_v2.Port.project_id)
+        query = query.join(ml2_models.PortBindingLevel, ml2_models.PortBindingLevel.port_id == models_v2.Port.id)
+        query = query.join(segment_models.NetworkSegment,
+                           ml2_models.PortBindingLevel.segment_id == segment_models.NetworkSegment.id)
+        query = query.filter(segment_models.NetworkSegment.physical_network.like('{}%'.format(physical_network_prefix)))
+        query = query.distinct()
+
+        return [entry.project_id for entry in query.all()]
+
+    def get_trunk_vlan_usage_on_project(self, context, project_id, segmentation_id=None):
+        # return vlan --> networks mapping for aci trunk ports inside a project
+        query = context.session.query(trunk_models.SubPort.segmentation_id, models_v2.Port.network_id)
+        query = query.filter(models_v2.Port.project_id == project_id)
+        query = query.join(trunk_models.SubPort, trunk_models.SubPort.port_id == models_v2.Port.id)
+        query = query.join(ml2_models.PortBinding, ml2_models.PortBinding.port_id == models_v2.Port.id)
+        query = query.filter(ml2_models.PortBinding.vif_type == aci_const.VIF_TYPE_ACI)
+        if segmentation_id:
+            query = query.filter(trunk_models.SubPort.segmentation_id == segmentation_id)
+        query = query.distinct()
+
+        # we only expect there to be one entry per segmentation id, but keep a list in case something goes wrong
+        vlan_map = {}
+        for entry in query.all():
+            vlan_map.setdefault(entry[0], set())
+            vlan_map[entry[0]].add(entry[1])
```

Comment:
With `setdefault()` returning what it found/set, you could put this into one line:
```python
vlan_map.setdefault(entry[0], set()).add(entry[1])
```